internal/slave/service.go

```diff
@@ -197,3 +238,138 @@ func newErrorStatus(err error) *serverpb.Status {
 func newEmptyStatus() *serverpb.Status {
 	return &serverpb.Status{Code: 0, Message: ""}
 }
+
+func (ss *slaveService) GetStatus(context context.Context, request *emptypb.Empty) (*serverpb.RegionInfo, error) {
+	if ss.replLag > ss.replConfig.MaxActiveReplLag {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else if ss.lastReplTime == 0 || hlc.GetTimeAgo(ss.lastReplTime) > ss.replConfig.MaxActiveReplElapsed {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else if ss.isClosed {
+		ss.regionInfo.Status = serverpb.RegionStatus_INACTIVE
+	} else {
+		ss.regionInfo.Status = serverpb.RegionStatus_ACTIVE_SLAVE
+	}
+	ss.regionInfo.MasterHost = &ss.replConfig.ReplMasterAddr
+	ss.lg.Debug("Current Info", zap.String("Status", ss.regionInfo.Status.String()),
+		zap.Uint64("Repl Lag", ss.replLag), zap.Uint64("Last Repl time", ss.lastReplTime))
+	return ss.regionInfo, nil
+}
+
+func (ss *slaveService) replaceMasterIfInactive() error {
+	if ss.replConfig.DisableAutoMasterDisc {
+		return nil
+	}
+	if regions, err := ss.clusterInfo.GetClusterStatus(ss.regionInfo.GetDatabase(), ss.regionInfo.GetVBucket()); err == nil {
+		currentMasterIdx := len(regions)
+		for i, region := range regions {
+			if region.NodeAddress == ss.replConfig.ReplMasterAddr {
+				currentMasterIdx = i
+				break
+			}
+		}
+
+		if currentMasterIdx == len(regions) {
```

Comment:
Why not simply do a findItem instead of comparing length?

Reply:
couldn't find any such API. this was the simplest implementation to check if value doesn't exist and if it exists to return the struct

Reply:
A simpler version would be

```go
var currentMaster *serverpb.RegionInfo = nil
for _, region := range regions {
	if region.NodeAddress == ss.replConfig.ReplMasterAddr {
		currentMaster = region
		break
	}
}

if currentMaster == nil {
	// Current Master not found in cluster. implies current master is inactive
	return ss.reconnectMaster()
} else if !isVBucketApplicableToBeMaster(currentMaster) {
	return ss.reconnectMaster()
} else {
	// current master is active. No action required as replication error could be temporary
	// need to validate this assumption though
	return nil
}
```

Reply:
done. but not much difference in code length. agree though better than len check